go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -548,40 +627,41 @@ func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
 	}
 
 	klog.V(5).Infof("Adding endpoints %s in namespace %s", ep.Name, ep.Namespace)
-	etpHostRules = hasHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses)
+	hasHostNetEndPnt := hasLocalHostNetworkEndpoints(ep, &npw.nodeIPManager.addresses)
+	hasNonHostNetEndPnt := (len(ep.Subsets) > 0 && !hasHostNetEndPnt)
 
 	// Here we make sure the correct rules are programmed whenever an AddEndpoint
 	// event is received, only alter flows if we need to, i.e if cache wasn't
-	// set or if it was and etpHostRules state changed, to prevent flow churn
-	out, exists := npw.getAndSetServiceInfo(name, svc, etpHostRules)
+	// set or if it was and svcHasLocalHostNetEndPnt state changed, to prevent flow churn
+	out, exists := npw.getAndSetServiceInfo(name, svc, hasHostNetEndPnt, hasNonHostNetEndPnt)
 	if !exists {
 		klog.V(5).Infof("Endpoint %s ADD event in namespace %s is creating rules", ep.Name, ep.Namespace)
-		addServiceRules(svc, etpHostRules, npw)
+		addServiceRules(svc, hasHostNetEndPnt, hasNonHostNetEndPnt, npw)
 		return
 	}
 
-	if out.etpHostRules != etpHostRules {
+	if out.etpHostRules != hasHostNetEndPnt || out.etpNonHostRules != hasNonHostNetEndPnt {
 		klog.V(5).Infof("Endpoint %s ADD event in namespace %s is updating rules", ep.Name, ep.Namespace)
 		delServiceRules(svc, npw)
-		addServiceRules(svc, etpHostRules, npw)
+		addServiceRules(svc, hasHostNetEndPnt, hasNonHostNetEndPnt, npw)
 	}
 
 }
 
 func (npw *nodePortWatcher) DeleteEndpoints(ep *kapi.Endpoints) {
-	var etpHostRules = false
+	var svcHasLocalHostNetEndPnt = false
+	var svcHasNonLocalHostNetEndPnt = false
 
 	klog.V(5).Infof("Deleting endpoints %s in namespace %s", ep.Name, ep.Namespace)
 	// remove rules for endpoints and add back normal ones
 	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
-	if svcConfig, exists := npw.updateServiceInfo(name, nil, &etpHostRules); exists {
+	if svcConfig, exists := npw.updateServiceInfo(name, nil, &svcHasLocalHostNetEndPnt, &svcHasNonLocalHostNetEndPnt); exists {
 		// Lock the cache mutex here so we don't miss a service delete during an endpoint delete
 		// we have to do this because deleting and adding iptables rules is slow.
 		npw.serviceInfoLock.Lock()
 		defer npw.serviceInfoLock.Unlock()
 
 		delServiceRules(svcConfig.service, npw)
-		addServiceRules(svcConfig.service, etpHostRules, npw)
```

Comment:
@astoycos, @trozet : Any idea why we were adding back the rules for endpoint deletion? Didn't make sense..so I removed it, let me know if there was some logic on why we did this.

Reply:
@tssurya We still need this because of two reasons I can think of right now....The Service rules (reguardless of `etpHostEndpoints` state) should only be fully deleted once the service is deleted, not upon endpoint deletion 
If the service exists without endpoints (weird scenario) we should still funnel into OVN to reject.

Reply:
ah ok, we just want to let the LB take care of all rejects if endpoints are not there right?